client/beefy/src/import.rs

```diff
@@ -0,0 +1,141 @@
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::{collections::HashMap, marker::PhantomData, sync::Arc};
+
+use sp_api::{ProvideRuntimeApi, TransactionFor};
+use sp_blockchain::well_known_cache_keys;
+use sp_consensus::Error as ConsensusError;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header as HeaderT},
+};
+
+use sc_client_api::backend::Backend;
+use sc_consensus::{BlockCheckParams, BlockImport, BlockImportParams, ImportResult};
+
+use crate::{
+	justification::decode_and_verify_commitment, notification::BeefySignedCommitmentSender,
+	Client as BeefyClient,
+};
+use beefy_primitives::{BeefyApi, VersionedFinalityProof, BEEFY_ENGINE_ID};
+
+/// A block-import handler for BEEFY.
+///
+/// This scans each imported block for BEEFY justifications and verifies them.
+/// Wraps a `inner: BlockImport` and ultimately defers to it.
+///
+/// When using BEEFY, the block import worker should be using this block import object.
+pub struct BeefyBlockImport<Backend, Block: BlockT, Client, I> {
+	client: Arc<Client>,
+	inner: I,
+	justification_sender: BeefySignedCommitmentSender<Block>,
+	_phantom: PhantomData<Backend>,
+}
+
+impl<BE, Block: BlockT, Client, I: Clone> Clone for BeefyBlockImport<BE, Block, Client, I> {
+	fn clone(&self) -> Self {
+		BeefyBlockImport {
+			client: self.client.clone(),
+			inner: self.inner.clone(),
+			justification_sender: self.justification_sender.clone(),
+			_phantom: PhantomData,
+		}
+	}
+}
+
+impl<BE, Block: BlockT, Client, I> BeefyBlockImport<BE, Block, Client, I> {
+	/// Create a new BeefyBlockImport.
+	pub fn new(
+		client: Arc<Client>,
+		inner: I,
+		justification_sender: BeefySignedCommitmentSender<Block>,
+	) -> BeefyBlockImport<BE, Block, Client, I> {
+		BeefyBlockImport { inner, client, justification_sender, _phantom: PhantomData }
+	}
+}
+
+#[async_trait::async_trait]
+impl<BE, Block: BlockT, Client, I> BlockImport<Block> for BeefyBlockImport<BE, Block, Client, I>
+where
+	BE: Backend<Block>,
+	I: BlockImport<
+			Block,
+			Error = ConsensusError,
+			Transaction = sp_api::TransactionFor<Client, Block>,
+		> + Send
+		+ Sync,
+	Client: BeefyClient<Block, BE> + ProvideRuntimeApi<Block>,
+	Client::Api: BeefyApi<Block>,
+{
+	type Error = ConsensusError;
+	type Transaction = TransactionFor<Client, Block>;
+
+	async fn import_block(
+		&mut self,
+		block: BlockImportParams<Block, Self::Transaction>,
+		new_cache: HashMap<well_known_cache_keys::Id, Vec<u8>>,
+	) -> Result<ImportResult, Self::Error> {
+		let hash = block.post_hash();
+		let number = *block.header.number();
+		let beefy_justification = block
+			.justifications
+			.as_ref()
+			.and_then(|just| just.get(BEEFY_ENGINE_ID).cloned());
+
+		// Run inner block import.
+		let import_result = self.inner.import_block(block, new_cache).await?;
```

Comment:
Yes, `inner` here will be (or in the future include) grandpa block import.

> it won't pass BEEFY justification to the client => it won't be stored in the database

I think you're right, I didn't realize the client will need it as well for storing in the db.
I cloned the beefy just before passing it down the line exactly because I saw that grandpa can take/consume all of them, so the `BeefyWorker` does get the justification, but the client backend+db might not.

Let's see what @andresilva thinks here before I start hacking the grandpa code.

> if justification is supposed to be saved in this call, then what if it is invalid?

Also good catch! I guess as long as we filter before BeefyWorker it still works, but filtering out invalid justif early avoids us caching and forwarding invalid justifications. I will fix this.

Reply:
because we don't want to tie BEEFY with GRANDPA logic:
1. the BEEFY justification is removed from list going downstream,
2. once all `inner` finish successfully, justification is _manually_ appended to `backend` if it is valid.